### PR TITLE
Control promotion

### PR DESCRIPTION
**Pull request by paili0628**
Expands the static island by aligning the separate static islands in `static seq` and `static par`
 For example:
 ```
 wires {
  static group g1<1> {...}
  static group g2<1> {...}
  static group g3<1> {...}
 }

 control { seq {g1; g2; g3;} }
 ```
 gets turned into
 ```
 wires {...}
 control { static seq {g1; g2; g3;} }
 ```
 while

 ```
 wires {
  static group g1<1> {...}
  static group g2<1> {...}
  group g3 {...}
 }

 control { seq {g1; g2; g3;} }
 ```
 gets turned into
 ```
 wires {...}
 control { seq {static seq {g1; g2;} g3;}}
 ```
 similarly for `static par`

### CODE REVIEW

**Comment by rachitnigam** [comment]
Looks cool! I'd recommend separating out the two things that this pass does:
1. Propagate `static` information from groups into control operators
2. Split up `seq` so that we get more `static seq`

Specifically, we want (1) for all control operators–when you have an `if` with only `static` groups, it better be turned into a `static if`. On the other hand (2) should probably be somewhat heuristic driven? It's not clear to me that creating static groups of size two is that much of a win but it's much more of a win to make more `static par` groups. In the paper, we should study the effect of various heuristics applied to these passes.

**Comment by sampsyo** [comment]
Really really nice!!! This seems like a very useful thing to have and has a clear definition of correctness. It seems nice for frontends to, on the whole, forget about whether they need to use static or dynamic control statements and mostly let this inference take care of it.

@rachitnigam makes an interesting point about somehow separating out the two kinds of change, perhaps into two completely different passes. To recap the reason(s) why this might be a good idea:

* As usual, it keeps passes focused and simple. Something that *just* inserts `static` in strategic locations in the control program is very easy to visually spot-check; the deeper rearrangement of `seq`s and stuff is a bit more complicated, so it's nice to have it in a separately-testable unit.
* One part is a clear, unambiguous win (there is no way it's ever a bad idea to add `static` to a control statement that can use it) and the other is more of a heuristic (there is kind of an implied cost trade-off to adding a nesting level to a `seq`, so it could be a win in some cases but not others). So it's nice to be able to disable the latter separately from the former.

However, I think it's tricky to think about how this would work, mechanically. There are a few ways I can imagine it, but a clear-ish one would be that they always happen in this order:

1. infer extra `static` annotations
2. opportunistically break up `seq`s

That is, stage 2 would use the results of stage 1. This means that if you have a control program that starts out like this:

```
seq {
  if ... { g1 };
  g2;
  g3;
  g4;
}
```

And then, after phase 1, you have:

```
seq {
  static if ... { g1 };
  g2;
  g3;
  g4;
}
```

…where `g2` is a static group but the last couple of groups are not, then phase 2 can still do its thing to that program and accomplish this:

```
seq {
  static seq {
    static if ... { g1 };
    g2;
  };
  g3;
  g4;
}
```

In other words, phase 2 can treat `static` control statements interchangeably, including enables.
